Client side profanity filter

edg-l opened the pull request:
Implements/fixes #2304

Although this version doesn't do confusables matching, but I leave it for the future.

I need to be given an official ddnet url to set as the default option  for the world list url before merging this.

![image](https://github.com/user-attachments/assets/355916de-c6f6-41ef-a21d-9c80dd48f68b)

![image](https://github.com/user-attachments/assets/f18b653a-8945-4dab-8e66-fd4e53f3e89c)


## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

heinrich5991 commented on src/game/client/components/menus_settings.cpp:
```diff
@@ -2415,6 +2415,16 @@ void CMenus::RenderSettingsAppearance(CUIRect MainView)
+
+		LeftView.HSplitTop(LineSize, &Button, &LeftView);
+		static int s_CensorChatOnline = 0;
+		if(DoButton_CheckBox(&s_CensorChatOnline, Localize("Censor profanity (Online & Local List)"), g_Config.m_ClCensorChat == 1, &Button))
```
```suggestion
		if(DoButton_CheckBox(&s_CensorChatOnline, Localize("Censor profanity (online & local list)"), g_Config.m_ClCensorChat == 1, &Button))
```

edg-l commented:
Should i implement this using std::set? would that be faster? (cost of hashing, etc)

edg-l commented:
> Would just the remote list suffice for now?
I think some people wouldnt enable the filter if they think the online list is too aggressive and just want some words they consider bad, and since its implemented already i dont see a problem

Chairn commented on src/game/client/components/chat.cpp:
```diff
@@ -532,6 +540,47 @@ void CChat::DisableMode()
+	if(g_Config.m_ClCensorChat == 0)
+		return;
+
+	if(g_Config.m_ClCensorChat == 1 || g_Config.m_ClCensorChat == 2)
```
useless if, config already guarentees it's either 1 or 2

Chairn commented on src/game/client/components/chat.cpp:
```diff
@@ -540,7 +589,16 @@ void CChat::OnMessage(int MsgType, void *pRawMsg)
+		if(g_Config.m_ClCensorChat != 0)
+		{
+			char aMessage[MAX_LINE_LENGTH];
+			str_copy(aMessage, pMsg->m_pMessage);
```
isn't it possible to do it directly in place ?

edg-l replied:
m_pMessage is const

heinrich5991 commented on src/game/client/components/chat.cpp:
```diff
@@ -1385,3 +1443,148 @@ void CChat::SendChatQueued(const char *pLine)
+	if(GotFileTimes)
+	{
+		pGet->IfModifiedSince(FileModifiedTime);
+		pGet->FailOnErrorStatus(false);
```
I'm gonna do a PR to implement this.

edg-l commented:
@heinrich5991 fixed and now uses json

```json
{
  "words": [
    "bad",
    "words"
  ]
}
```

heinrich5991 commented:
I added a sample censor.json to https://info.ddnet.org/censor.json, I guess you can point it there.

def- commented:
Fine for me.

edg-l commented:
> I added a sample censor.json to https://info.ddnet.org/censor.json, I guess you can point it there.
> 
> Is that a correct place for it? @def- @Robyt3?

updated

Robyt3 commented on src/game/client/components/chat.cpp:
```diff
@@ -1385,3 +1459,144 @@ void CChat::SendChatQueued(const char *pLine)
+	// Load existing file while waiting for the HTTP request
+	if(GotFileTimes)
+	{
+		m_pChat->m_vCensoredWords.clear();
+		m_pChat->LoadCensorList(m_aSaveFilePath, true);
```
This shouldn't modify the chat/censor component's block list directly, that's probably not thread-safe. Generally the way our jobs work is that they store their result locally and the main threads polls the job and gets the results from the job when it's done. So the `LoadCensorList` function should return a vector which is stored in the job until it's fetched by the main thread (with `std::move`).

edg-l commented:
@Robyt3 fixed all i think

heinrich5991 commented on src/game/client/components/censor.cpp:
```diff
@@ -0,0 +1,238 @@
+			pStart = str_utf8_find_nocase(pStart, Word.c_str(), &pEnd);
+			if(!pStart)
+				continue;
+			if((pStart == pBuffer || str_utf8_isspace(*(pStart - 1))) && (str_utf8_isspace(*(pEnd))))
```
I guess it would be interesting to hear from a stakeholder whether this does enough to be worth including: @murpii Do you think a words-based blacklist is useful enough to be included? You'd be able to block words like typed `hello badbadbad`, but copypasted (containing a cyrillic а) `hello bаdbadbad` would not get caught, neither would `hello badbadbad!` or `hello .badbadbad`.

murpii replied:
> I guess it would be interesting to hear from a stakeholder whether this does enough to be worth including: @murpii Do you think a words-based blacklist is useful enough to be included? You'd be able to block words like typed `hello badbadbad`, but copypasted (containing a cyrillic а) `hello bаdbadbad` would not get caught, neither would `hello badbadbad!` or `hello .badbadbad`.

Weird, didn't get notified by the @mention. I think you're better off asking our moderators. ^^

heinrich5991 replied:
@bencie You reacted with a ❤️, and you might also be someone who uses this. Can you comment on whether this blacklist would be useful to you?

Learath2 replied:
Streamers have been requesting this for a while, too. I think it's much more useful to them than the average user. Maybe ask lcsg?

bencie replied:
> @bencie You reacted with a ❤️, and you might also be someone who uses this. Can you comment on whether this blacklist would be useful to you?

Yes, for me, it'd mainly be useful against annoying chat spam, like huge ascii binds. My current method is to just add_foe everyone who uses a bind that fills your chat like that. It's highly inefficient because I'd stlll prefer to be able to see their normal messages. And as Learath said, streamers would be using this feature a lot too.

heinrich5991 replied:
> And as Learath said, streamers would be using this feature a lot too.

I'm not really asking whether chat censoring is useful to you or to streamers. I want to ask whether **this specific method** of censoring chat is useful to you. I tried to outline its limitations above:

> Do you think a words-based blacklist is useful enough to be included? You'd be able to block words like typed `hello badbadbad`, but copypasted (containing a cyrillic а) `hello bаdbadbad` would not get caught, neither would `hello badbadbad!` or hello `.badbadbad`.

---

> Yes, for me, it'd mainly be useful against annoying chat spam, like huge ascii binds.

I'm not sure if this is effective. The huge chat bind would be replaced by lots of stars, so it would still spam the chat. Since I think you've misunderstood how this works, I'd like to ask you again if **this particular implementation** is useful to you — I understand that the general idea of chat censoring is useful, but I'm trying to drill down what exactly we need.

bencie replied:
In it's current state, I would not find it useful for me specifically, no.

Robyt3 commented on src/game/client/components/censor.cpp:
```diff
@@ -0,0 +1,238 @@
+		return;
+	}
+
+	log_info("chat-censor", "downloaded word censor blacklist succesfully");
```
`successfully`

edg-l commented:
![image](https://github.com/user-attachments/assets/fd572404-31b6-4006-bf9f-72921dc030ea)

@Robyt3 updated

Robyt3 commented on src/game/client/components/censor.cpp:
```diff
@@ -0,0 +1,237 @@
+	return true;
+}
+
+void CCensor::CBlacklistDownloadJob::Run()
```
This should be simpler now with #9574.

Robyt3 commented on src/game/client/components/menus_settings.cpp:
```diff
@@ -2422,6 +2422,21 @@ void CMenus::RenderSettingsAppearance(CUIRect MainView)
 
+		// Chat censor selection
+		CUIRect CensorModeDropDown;
+		LeftView.HSplitTop(20.0f, &CensorModeDropDown, &LeftView);
```
Also add a small spacing above and below the dropdown button: `LeftView.HSplitTop(2.0f, nullptr, &LeftView);`

Robyt3 commented on src/game/client/components/censor.h:
```diff
@@ -0,0 +1,57 @@
+	class CBlacklistDownloadJob : public IJob
+	{
+	public:
+		CBlacklistDownloadJob(CCensor *pCensor, const char *pURL, const char *pSaveFilePath);
```
`pUrl`

Robyt3 commented on src/game/client/components/censor.cpp:
```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+void CCensor::CensorMessage(char *pMessage)
```
This function can be `const`

Robyt3 commented on src/game/client/components/censor.cpp:
```diff
@@ -0,0 +1,237 @@
+	unsigned FileSize;
+	if(!Storage()->ReadFile(pFilePath, IStorage::TYPE_SAVE, &pFileData, &FileSize))
+	{
+		log_error("chat-censor", "Failed to open/read word censor file '%s'", pFilePath);
```
Change to log system `"censor"`?

edg-l commented:
@Robyt3 is it ok to go?

heinrich5991 commented:
I'll let someone else chime in on this, but I think supporting local files as well brings unnecessary complexity and could simply be removed — this way, we wouldn't have to maintain it. @Robyt3 What do you think?

I think the implementation is a good first step towards an implementation of #2304 and should be merged as a first step. It doesn't really address the issue #2304 though, so the PR shouldn't close it. #2304 needs a more comprehensive word filtering that is not trivially sidesteppable, e.g. by using confusables data.

Robyt3 reviewed (commented):
> I'll let someone else chime in on this, but I think supporting local files as well brings unnecessary complexity and could simply be removed — this way, we wouldn't have to maintain it. @Robyt3 What do you think?

Would be fine with me to keep it simpler. In the Discord thread there were only very few comments explicitly in favor of a separate local list, e.g.

> +1 for me, being able to enable/disable and add certain words seems great - @BlaiZephyr

Which wouldn't be solved by a separate local list alone either as it doesn't easily allow to enable/disable existing words.

I think only supporting an online list would also be enough. Those who want to override the online list can host their own list (locally) or replace the local file and set it to read-only.

> I think the implementation is a good first step towards an implementation of #2304 and should be merged as a first step. It doesn't really address the issue #2304 though, so the PR shouldn't close it. #2304 needs a more comprehensive word filtering that is not trivially sidesteppable, e.g. by using confusables data.

Yeah, @edg-l please update the description so the issue is not closed automatically.

edg-l commented:
Removed the local list and fixed conflicts

Robyt3 reviewed (approved):
LGTM as a first step. Please squash the commits.

ChillerDragon commented:
<img width="443" height="114" alt="image" src="https://github.com/user-attachments/assets/9d8fa7af-422e-4b4b-a6b8-ccd6007a7848" />
